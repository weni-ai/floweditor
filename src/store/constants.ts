enum Constants {
  UPDATE_EDITOR_STATE = 'UPDATE_EDITOR_STATE',
  UPDATE_NODE_EDITOR_SETTINGS = 'UPDATE_NODE_EDITOR_SETTINGS',
  UPDATE_TRANSLATING = 'UPDATE_TRANSLATING',
  UPDATE_BASE_LANGUAGE = 'UPDATE_BASE_LANGUAGE',
  UPDATE_SEARCH = 'UPDATE_SEARCH',
  UPDATE_LANGUAGES = 'UPDATE_LANGUAGES',
  UPDATE_LANGUAGE = 'UPDATE_LANGUAGE',
  UPDATE_FETCHING_FLOW = 'FETCHING_FLOW',
  UPDATE_DEFINITION = 'UPDATE_DEFINITION',
  UPDATE_NODE_DRAGGING = 'UPDATE_NODE_DRAGGING',
  UPDATE_FLOWS = 'UPDATE_FLOWS',
  UPDATE_PENDING_CONNECTIONS = 'UPDATE_PENDING_CONNECTIONS',
  REMOVE_PENDING_CONNECTION = 'REMOVE_PENDING_CONNECTION',
  UPDATE_COMPONENTS = 'UPDATE_COMPONENTS',
  UPDATE_CONTACT_FIELDS = 'UPDATE_CONTACT_FIELDS',
  UPDATE_GROUPS = 'UPDATE_GROUPS',
  UPDATE_RESULT_MAP = 'UPDATE_RESULT_MAP',
  UPDATE_ASSET_MAP = 'UPDATE_ASSET_MAP',
  UPDATE_NODES = 'UPDATE_NODES',
  UPDATE_ISSUES = 'UPDATE_ISSUES',
  UPDATE_METADATA = 'UPDATE_METADATA',
  UPDATE_EXITS = 'UPDATE_EXITS',
  UPDATE_NODE_EDITOR_OPEN = 'UPDATE_NODE_EDITOR_OPEN',
  UPDATE_CREATE_NODE_POSITION = 'UPDATE_CREATE_NODE_POSITION',
  UPDATE_DRAG_GROUP = 'UPDATE_DRAG_GROUP',
  UPDATE_TYPE_CONFIG = 'UPDATE_TYPE_CONFIG',
  UPDATE_USER_ADDING_ACTION = 'UPDATE_USER_ADDING_ACTION',
  UPDATE_SEARCH_BAR = 'UPDATE_SEARCH_BAR',
  INCREMENT_SUGGESTED_RESULT_NAME_COUNT = 'INCREMENT_SUGGESTED_RESULT_NAME_COUNT'
}

export default Constants;
