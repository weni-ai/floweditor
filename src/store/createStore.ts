import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  return store;
};
