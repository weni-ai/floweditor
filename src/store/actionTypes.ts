import { Type } from 'config/interfaces';
import { FlowDefinition, FlowMetadata } from 'flowTypes';
import Constants from 'store/constants';
import { EditorState } from 'store/editor';
import {
  Asset,
  AssetStore,
  ContactFields,
  RenderNodeMap,
  FlowIssueMap,
  Search
} from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

// Redux action generic
interface DuxAction<T extends Constants, P extends { [key: string]: any } = {}> {
  type: T;
  payload?: P;
}

// Payload types
interface EditorStatePayload {
  editorState: EditorState;
}

interface DefinitionPayload {
  definition: FlowDefinition;
}

interface BaseLanguagePayload {
  baseLanguage: Asset;
}

interface SearchPayload {
  search: Search;
}

interface UpdateMetadataPayload {
  metadata: FlowMetadata;
}

interface LanguagesPayload {
  languages: Asset[];
}

interface UpdateFlowsPayload {
  flows: { uuid: string; name: string }[];
}

interface UpdateDependenciesPayload {
  dependencies: FlowDefinition[];
}

interface UpdateAssetMapPayload {
  assets: AssetStore;
}

interface UpdateNodesPayload {
  nodes: RenderNodeMap;
}

interface UpdateIssuesPayload {
  issues: FlowIssueMap;
}

interface UpdateTypeConfigPayload {
  typeConfig: Type;
}

interface UpdateUserAddingActionPayload {
  userAddingAction: boolean;
}

interface UpdateNodeEditorSettingsPayload {
  settings: NodeEditorSettings;
}

interface UpdateContactFieldsPayload {
  contactFields: ContactFields;
}

// Action types
export type UpdateNodeEditorSettings = DuxAction<
  Constants.UPDATE_NODE_EDITOR_SETTINGS,
  UpdateNodeEditorSettingsPayload
>;

export type UpdateEditorState = DuxAction<Constants.UPDATE_EDITOR_STATE, EditorStatePayload>;

export type UpdateBaseLanguageAction = DuxAction<
  Constants.UPDATE_BASE_LANGUAGE,
  BaseLanguagePayload
>;

export type UpdateSearchAction = DuxAction<Constants.UPDATE_SEARCH, SearchPayload>;

export type UpdateLanguagesAction = DuxAction<Constants.UPDATE_LANGUAGES, LanguagesPayload>;

export type UpdateDefinitionAction = DuxAction<Constants.UPDATE_DEFINITION, DefinitionPayload>;

export type UpdateFlowsAction = DuxAction<Constants.UPDATE_FLOWS, UpdateFlowsPayload>;

export type UpdateMetadataAction = DuxAction<Constants.UPDATE_METADATA, UpdateMetadataPayload>;

export type UpdateAssetsAction = DuxAction<Constants.UPDATE_ASSET_MAP, UpdateAssetMapPayload>;

export type IncrementSuggestedResultNameCountAction = DuxAction<
  Constants.INCREMENT_SUGGESTED_RESULT_NAME_COUNT
>;

export type UpdateNodesAction = DuxAction<Constants.UPDATE_NODES, UpdateNodesPayload>;

export type UpdateIssuesAction = DuxAction<Constants.UPDATE_ISSUES, UpdateIssuesPayload>;

export type UpdateTypeConfigAction = DuxAction<
  Constants.UPDATE_TYPE_CONFIG,
  UpdateTypeConfigPayload
>;

export type UpdateUserAddingActionAction = DuxAction<
  Constants.UPDATE_USER_ADDING_ACTION,
  UpdateUserAddingActionPayload
>;

export type UpdateContactFieldsAction = DuxAction<
  Constants.UPDATE_CONTACT_FIELDS,
  UpdateContactFieldsPayload
>;

export type UpdateTypeConfig = (typeConfig: Type) => UpdateTypeConfigAction;

export type UpdateUserAddingAction = (userAddingAction: boolean) => UpdateUserAddingActionAction;

export type UpdateBaseLanguage = (baseLanguage: Asset) => UpdateBaseLanguageAction;

export type UpdateSearch = (search: Search) => UpdateSearch;

export type IncrementSuggestedResultNameCount = () => IncrementSuggestedResultNameCountAction;

type ActionTypes =
  | UpdateEditorState
  | UpdateNodeEditorSettings
  | UpdateDefinitionAction
  | UpdateFlowsAction
  | UpdateMetadataAction
  | UpdateAssetsAction
  | IncrementSuggestedResultNameCountAction
  | UpdateNodesAction
  | UpdateIssuesAction
  | UpdateTypeConfigAction
  | UpdateUserAddingActionAction
  | UpdateBaseLanguageAction
  | UpdateLanguagesAction
  | UpdateContactFieldsAction
  | UpdateSearchAction;

export default ActionTypes;
