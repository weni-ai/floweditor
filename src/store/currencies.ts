import { AssetType } from 'store/flowContext';

export const currencies = {
  ARS: {
    id: 'ARS',
    name: 'Argentine Peso',
    type: AssetType.Currency,
  },
  AUD: {
    id: 'AUD',
    name: 'Australian Dollar',
    type: AssetType.Currency,
  },
  BIF: {
    id: 'BIF',
    name: 'Burundi Franc',
    type: AssetType.Currency,
  },
  BRL: {
    id: 'BRL',
    name: 'Brazilian Real',
    type: AssetType.Currency,
  },

  CAD: {
    id: 'CAD',
    name: 'Canadian Dollar',
    type: AssetType.Currency,
  },
  CDF: {
    id: 'CDF',
    name: 'Congolese Franc',
    type: AssetType.Currency,
  },
  CLP: {
    id: 'CLP',
    name: 'Chilean Peso',
    type: AssetType.Currency,
  },
  COP: {
    id: 'COP',
    name: 'Colombian Peso',
    type: AssetType.Currency,
  },
  DJF: {
    id: 'DJF',
    name: 'Djibouti Franc',
    type: AssetType.Currency,
  },
  DOP: {
    id: 'DOP',
    name: 'Dominican Peso',
    type: AssetType.Currency,
  },
  DZD: {
    id: 'DZD',
    name: 'Algerian Dinar',
    type: AssetType.Currency,
  },
  EUR: {
    id: 'EUR',
    name: 'Euro',
    type: AssetType.Currency,
  },
  GBP: {
    id: 'GBP',
    name: 'Pound Sterling',
    type: AssetType.Currency,
  },
  GHS: {
    id: 'GHS',
    name: 'Ghana Cedi',
    type: AssetType.Currency,
  },
  GNF: {
    id: 'GNF',
    name: 'Guinean Franc',
    type: AssetType.Currency,
  },
  KES: {
    id: 'KES',
    name: 'Kenyan Shilling',
    type: AssetType.Currency,
  },
  LBP: {
    id: 'LBP',
    name: 'Lebanese Pound',
    type: AssetType.Currency,
  },
  LKR: {
    id: 'LKR',
    name: 'Sri Lanka Rupee',
    type: AssetType.Currency,
  },
  LRD: {
    id: 'LRD',
    name: 'Liberian Dollar',
    type: AssetType.Currency,
  },
  MWK: {
    id: 'MWK',
    name: 'Malawi Kwacha',
    type: AssetType.Currency,
  },
  MXN: {
    id: 'MXN',
    name: 'Mexican Peso',
    type: AssetType.Currency,
  },
  NAD: {
    id: 'NAD',
    name: 'Namibia Dollar',
    type: AssetType.Currency,
  },
  NPR: {
    id: 'NPR',
    name: 'Nepalese Rupee',
    type: AssetType.Currency,
  },
  NZD: {
    id: 'NZD',
    name: 'New Zealand Dollar',
    type: AssetType.Currency,
  },
  PHP: {
    id: 'PHP',
    name: 'Philippine Piso',
    type: AssetType.Currency,
  },
  PKR: {
    id: 'PKR',
    name: 'Pakistan Rupee',
    type: AssetType.Currency,
  },
  RUB: {
    id: 'RUB',
    name: 'Russian Ruble',
    type: AssetType.Currency,
  },
  RWF: {
    id: 'RWF',
    name: 'Rwanda Franc',
    type: AssetType.Currency,
  },
  SDG: {
    id: 'SDG',
    name: 'Sudanese Pound',
    type: AssetType.Currency,
  },
  SGD: {
    id: 'SGD',
    name: 'Singapore Dollar',
    type: AssetType.Currency,
  },
  SLL: {
    id: 'SLL',
    name: 'Leone',
    type: AssetType.Currency,
  },
  SOS: {
    id: 'SOS',
    name: 'Somali Shilling',
    type: AssetType.Currency,
  },
  SRD: {
    id: 'SRD',
    name: 'Surinam Dollar',
    type: AssetType.Currency,
  },
  SVC: {
    id: 'SVC',
    name: 'El Salvador Colon',
    type: AssetType.Currency,
  },
  TND: {
    id: 'TND',
    name: 'Tunisian Dinar',
    type: AssetType.Currency,
  },
  TRY: {
    id: 'TRY',
    name: 'Turkish Lira',
    type: AssetType.Currency,
  },
  TZS: {
    id: 'TZS',
    name: 'Tanzanian Shilling',
    type: AssetType.Currency,
  },
  UGX: {
    id: 'UGX',
    name: 'Uganda Shilling',
    type: AssetType.Currency,
  },
  USD: {
    id: 'USD',
    name: 'US Dollar',
    type: AssetType.Currency,
  },
  UYU: {
    id: 'UYU',
    name: 'Peso Uruguayo',
    type: AssetType.Currency,
  },
  UZS: {
    id: 'UZS',
    name: 'Uzbekistan Sum',
    type: AssetType.Currency,
  },
  VEF: {
    id: 'VEF',
    name: 'Bol√≠var',
    type: AssetType.Currency,
  },
  XCD: {
    id: 'XCD',
    name: 'East Caribbean Dollar',
    type: AssetType.Currency,
  },
  YER: {
    id: 'YER',
    name: 'Yemeni Rial',
    type: AssetType.Currency,
  },
  ZAR: {
    id: 'ZAR',
    name: 'Rand',
    type: AssetType.Currency,
  },
  ZMW: {
    id: 'ZMW',
    name: 'Zambian Kwacha',
    type: AssetType.Currency,
  },
  ZWL: {
    id: 'ZWL',
    name: 'Zimbabwe Dollar',
    type: AssetType.Currency,
  },
};
