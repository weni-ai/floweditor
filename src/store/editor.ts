import { FlowPosition } from 'flowTypes';
import ActionTypes, { UpdateEditorState } from 'store/actionTypes';
import Constants from 'store/constants';
import { Asset, RenderNode } from 'store/flowContext';

// tslint:disable:no-shadowed-variable
export interface DragSelection {
  startX?: number;
  startY?: number;
  currentX?: number;
  currentY?: number;
}

export interface DebugState {
  showUUIDs: boolean;
}

export interface CanvasPositions {
  [uuid: string]: FlowPosition;
}

export interface Activity {
  nodes: { [uuid: string]: number };
  segments: { [exitToNodeKey: string]: number };
  recentMessages?: { [key: string]: RecentMessage[] };
  is_starting?: boolean;
}

export interface RecentMessage {
  sent: string;
  text: string;
}

export interface Warning {
  name: string;
  type: string;
}

export interface Warnings {
  [uuid: string]: Warning;
}

export interface EditorState {
  currentRevision: number | null;
  simulating: boolean;
  language: Asset | null;
  translating: boolean;
  fetchingFlow: boolean;
  ghostNode: RenderNode | null;
  dragActive: boolean;
  dragStartTime: number;
  dragDownPosition: FlowPosition | null;
  dragNodeUUID: string | null;
  dragGroup: boolean;
  dragSelection: DragSelection | null;
  selectionActive: boolean;
  debug?: DebugState | null;

  // which poptab is popped
  popped: string;

  warnings: Warnings;

  modalMessage?: ModalMessage;
  saving?: boolean;

  // the currently shown activity, can be
  // simulation or live
  activity: Activity;

  // the current live activity
  liveActivity: Activity;

  // interval in millis we should refresh activity
  activityInterval: number;

  // is our page visible or tabbed away
  visible: boolean;

  scrollToNode: string;
  scrollToAction: string;

  currentGuide: string;
  guidingStep: number;

  mouseState: MouseState;
}

export interface ModalMessage {
  title: string;
  body: string;
}

export enum MouseState {
  SELECT = 'select',
  DRAG = 'drag',
  DRAGGING = 'dragging'
}

export const EMPTY_DRAG_STATE: any = {
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null
};

// Initial state
export const initialState: EditorState = {
  currentRevision: null,
  simulating: false,
  translating: false,
  language: null,
  fetchingFlow: false,
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null,
  ghostNode: null,
  selectionActive: false,
  debug: null,
  warnings: {},

  popped: null,

  activity: { segments: {}, nodes: {} },
  liveActivity: { segments: {}, nodes: {} },
  activityInterval: 5000,
  visible: true,

  scrollToNode: null,
  scrollToAction: null,

  currentGuide: null,
  guidingStep: -1,

  mouseState: MouseState.SELECT
};

// Action Creator
export const updateEditorState = (editorState: EditorState): UpdateEditorState => ({
  type: Constants.UPDATE_EDITOR_STATE,
  payload: {
    editorState
  }
});

/* Reducer */
export const editorState = (editorState: EditorState = initialState, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_EDITOR_STATE:
      return action.payload!.editorState;
    default:
      return editorState;
  }
};

// export our reducer
export default editorState;
