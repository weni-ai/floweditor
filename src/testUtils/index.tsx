// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
