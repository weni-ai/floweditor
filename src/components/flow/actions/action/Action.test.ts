import {
  actionContainerSpecId,
  actionInteractiveDivSpecId,
  ActionWrapper,
  ActionWrapperProps,
} from 'components/flow/actions/action/Action';
import { composeComponentTestUtils, getSpecWrapper, setMock } from 'testUtils';
import {
  createExit,
  createRenderNode,
  createSendMsgAction,
  createStartFlowAction,
  createSubflowNode,
  English,
  createSendWhatsAppMsgAction,
  createSendEmailAction,
  createCallWeniGPTAction,
  Spanish,
} from 'testUtils/assetCreators';
import { getLocalization, set, setFalse, setTrue } from 'utils';
import { MouseState } from 'store/editor';
import { shallowToJson } from 'enzyme-to-json';

const sendMsgAction = createSendMsgAction();
const sendMsgAction1 = createSendMsgAction({
  uuid: '0ed4b887-2924-45ea-babe-4d580cdfb000',
  text: 'Yo!',
});
const sendWppMsgAction = createSendWhatsAppMsgAction();
const sendEmailAction = createSendEmailAction();
const callWeniGptAction = createCallWeniGPTAction();
const sendMsgNode = createRenderNode({
  actions: [sendMsgAction],
  exits: [createExit()],
});
const startFlowAction = createStartFlowAction();
const subflowNode = createSubflowNode(startFlowAction);
const localization = {
  spa: {
    [sendMsgAction.uuid]: {
      text: ['¡Hola!'],
    },
  },
};

const wppLocalization = {
  spa: {
    [sendWppMsgAction.uuid]: {
      text: ['¡Hola!'],
    },
  },
};

const baseProps: ActionWrapperProps = {
  selected: false,
  localization,
  first: true,
  action: sendMsgAction,
  render: vi.fn(),
  renderNode: sendMsgNode,
  language: English,
  translating: false,
  onOpenNodeEditor: vi.fn(),
  removeAction: vi.fn(),
  moveActionUp: vi.fn(),
  issues: [],
  scrollToAction: '',
  mouseState: MouseState.SELECT,
};

const { setup, spyOn } = composeComponentTestUtils(ActionWrapper, baseProps);

describe(ActionWrapper.name, () => {
  describe('render', () => {
    it('should render self, children with base props', () => {
      const { wrapper, props, instance } = setup(true, {
        render: setMock(),
      });
      const actionContainer = getSpecWrapper(wrapper, actionContainerSpecId);

      expect(actionContainer.prop('id')).toBe(`action-${props.action.uuid}`);
      expect(
        getSpecWrapper(wrapper, actionInteractiveDivSpecId).exists(),
      ).toBeTruthy();
      expect(wrapper.find('TitleBar').props()).toMatchSnapshot();
      expect(props.render).toHaveBeenCalledTimes(1);
      expect(props.render).toHaveBeenCalledWith(
        props.action,
        instance.context.config.endpoints,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should show move icon', () => {
      const { wrapper } = setup(true, { first: setFalse() });

      expect(wrapper.find('TitleBar').prop('showMove')).toBeTruthy();
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display translating style', () => {
      const { wrapper } = setup(true, { translating: setTrue() });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should render selected', () => {
      const { wrapper } = setup(true, { selected: setTrue() });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display translating style to send whatsapp message action', () => {
      const { wrapper } = setup(true, {
        action: set(sendWppMsgAction),
        translating: setTrue(),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display translating style to send email action', () => {
      const { wrapper } = setup(true, {
        action: set(sendEmailAction),
        translating: setTrue(),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display translating style to call weni gpt action', () => {
      const { wrapper } = setup(true, {
        action: set(callWeniGptAction),
        translating: setTrue(),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display not_localizable style', () => {
      const { wrapper } = setup(true, {
        action: set(startFlowAction),
        translating: setTrue(),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display hybrid style', () => {
      const { wrapper } = setup(true, {
        renderNode: set(subflowNode),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should display missing_localization style', () => {
      const { wrapper } = setup(true, {
        action: set(sendMsgAction1),
        translating: setTrue(),
      });

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('instance methods', () => {
    describe('handleActionClicked', () => {
      it('should be called when interactive div is clicked', () => {
        const onClickSpy = spyOn('handleActionClicked');
        const { wrapper } = setup();
        const interactiveDiv = getSpecWrapper(
          wrapper,
          actionInteractiveDivSpecId,
        );
        const mockEvent = {
          preventDefault: vi.fn(),
          stopPropagation: vi.fn(),
        };

        interactiveDiv.simulate('mouseDown', mockEvent);
        interactiveDiv.simulate('mouseUp', mockEvent);

        expect(onClickSpy).toHaveBeenCalledTimes(1);

        onClickSpy.mockRestore();
      });

      it("should call 'onOpenEditor' action creator if node is not dragging", () => {
        const { props, instance } = setup(true, {
          onOpenNodeEditor: setMock(),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };

        const mockEvent: any = {
          preventDefault: vi.fn(),
          stopPropagation: vi.fn(),
        };

        instance.handleActionClicked(mockEvent);

        expect(props.onOpenNodeEditor).toHaveBeenCalledTimes(1);
        /* 
                expect(props.onOpenNodeEditor).toHaveBeenCalledWith({
                    originalNode: props.node,
                    originalAction: props.action,
                    showAdvanced: false
                });
                */
      });

      it("should call onOpenNodeEditor with showAdvanced true if clicked on a target with 'data-advanced' attribute", () => {
        const { props, instance } = setup(true, {
          onOpenNodeEditor: setMock(),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };

        const mockEvent: any = {
          preventDefault: vi.fn(),
          stopPropagation: vi.fn(),
          target: {
            attributes: {},
            getAttribute: vi.fn(() => 'true'),
          },
        };

        instance.handleActionClicked(mockEvent);

        expect(props.onOpenNodeEditor).toHaveBeenCalledTimes(1);
        expect(props.onOpenNodeEditor).toHaveBeenCalledWith({
          originalNode: props.renderNode,
          originalAction: props.action,
          showAdvanced: true,
        });
      });

      it('should not call onOpenNodeEditor if node is dragging', () => {
        const { props, instance } = setup(true, {
          onOpenNodeEditor: setMock(),
          mouseState: set(MouseState.DRAGGING),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };

        const mockEvent: any = {
          preventDefault: vi.fn(),
          stopPropagation: vi.fn(),
        };

        instance.handleActionClicked(mockEvent);

        expect(props.onOpenNodeEditor).not.toHaveBeenCalled();
      });
    });

    describe('handleRemoval', () => {
      it('should call removeAction action creator', () => {
        const { props, instance } = setup(true, {
          removeAction: setMock(),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };
        // const mockEvent: any = {
        //   stopPropagation: vi.fn(),
        //   preventDefault: vi.fn(),
        // };

        instance.handleRemoval();

        expect(props.removeAction).toHaveBeenCalledTimes(1);
        expect(props.removeAction).toHaveBeenCalledWith(
          props.renderNode.node.uuid,
          props.action,
        );
      });
    });

    describe('onMoveUp', () => {
      it('should call moveActionUp action creator', () => {
        const { props, instance } = setup(true, {
          moveActionUp: setMock(),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };

        const mockEvent: any = {
          stopPropagation: vi.fn(),
          preventDefault: vi.fn(),
        };

        instance.handleMoveUp(mockEvent);

        expect(props.moveActionUp).toHaveBeenCalledTimes(1);
        expect(props.moveActionUp).toHaveBeenCalledWith(
          props.renderNode.node.uuid,
          props.action,
        );
      });

      it('should call moveActionUp eve without event', () => {
        const { props, instance } = setup(true, {
          moveActionUp: setMock(),
        }) as { instance: ActionWrapper; props: ActionWrapperProps };

        instance.handleMoveUp(null);

        expect(props.moveActionUp).toHaveBeenCalledTimes(1);
        expect(props.moveActionUp).toHaveBeenCalledWith(
          props.renderNode.node.uuid,
          props.action,
        );
      });
    });

    describe('getAction', () => {
      it('should return the action passed via props if not localized', () => {
        const { props, instance } = setup(true, {
          node: set(sendMsgAction1),
        });

        expect(instance.getAction()).toEqual(props.action);
      });

      it('should return localized action if localized', () => {
        const { props, instance } = setup();
        const localizedObject = getLocalization(
          props.action,
          props.localization,
          props.language,
        ).getObject();

        expect(instance.getAction()).toEqual(localizedObject);
      });

      it('should return localized action if localized when translating', () => {
        const { props, instance } = setup(true, {
          action: set(sendWppMsgAction),
          translating: setTrue(),
          language: set(Spanish),
          localization: set(wppLocalization),
        });
        const localizedObject = getLocalization(
          props.action,
          props.localization,
          props.language,
        ).getObject();

        expect(instance.getAction()).toEqual(localizedObject);
      });
    });

    describe('scrollToAction', () => {
      it("should render MountScroll with action's uuid", () => {
        const { wrapper } = setup(true, {
          scrollToAction: set(sendMsgAction.uuid),
          action: set({ uuid: sendMsgAction.uuid }),
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('should render action titlebar with missing style if have any issues', () => {
      it('should render if have issues', () => {
        const { wrapper } = setup(true, {
          issues: set([
            {
              type: 'missing_dependency',
              node_uuid: sendMsgNode.node.uuid,
              action_uuid: sendMsgAction.uuid,
              description: 'Missing dependency',
            },
          ]),
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
    });
  });
});
