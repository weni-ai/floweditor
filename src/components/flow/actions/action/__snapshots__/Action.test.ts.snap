// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionWrapper > render > should display hybrid style 1`] = `ShallowWrapper {}`;

exports[`ActionWrapper > render > should display missing_localization style 1`] = `ShallowWrapper {}`;

exports[`ActionWrapper > render > should display not_localizable style 1`] = `ShallowWrapper {}`;

exports[`ActionWrapper > render > should display translating style 1`] = `ShallowWrapper {}`;

exports[`ActionWrapper > render > should render self, children with base props 1`] = `
{
  "__className": "_send_msg_9658e1",
  "hasIssues": false,
  "new": undefined,
  "onMoveUp": [Function],
  "onRemoval": [Function],
  "selected": false,
  "shouldCancelClick": [Function],
  "showMove": false,
  "showRemoval": true,
  "title": "Send message",
}
`;

exports[`ActionWrapper > render > should render self, children with base props 2`] = `ShallowWrapper {}`;

exports[`ActionWrapper > render > should show move icon 1`] = `ShallowWrapper {}`;
